File: src/services/deepseek.ts
import { AIService, AIServiceResult, GenerationOptions, AIMessage } from './types'
import { DesignJSON } from '@/types'
import { ERROR_MESSAGES } from '@/constants'
import {
  DEEPSEEK_STAGE_A_SYSTEM,
  DEEPSEEK_STAGE_B_SYSTEM,
  createStageAUserPrompt,
  createStageBUserPrompt,
} from '@/lib/prompts'

/**
 * DeepSeek AI服务实现类
 *
 * 使用DeepSeek API进行两阶段处理：
 * 1. 分析阶段：分析输入内容，生成设计JSON
 * 2. 渲染阶段：将设计JSON转换为SVG卡片
 *
 * @example
 * ```typescript
 * const service = new DeepSeekService(config)
 * const result = await service.process("文章内容")
 * console.log(result.svgContent, result.designJson)
 * ```
 */
export class DeepSeekService extends AIService {
  constructor(config: any) {
    super(config, 'DeepSeek')
  }

  /**
   * 处理文本内容，生成小红书卡片
   *
   * @param text 输入文本
   * @param options 生成选项
   * @returns Promise<AIServiceResult> 处理结果
   */
  async process(text: string, options?: GenerationOptions): Promise<AIServiceResult> {
    this.logInfo('开始处理文本', { textLength: text.length, options })

    try {
      // 阶段A：分析与设计JSON生成
      const designJson = await this.executeStageA(text, options)

      // 阶段B：SVG渲染
      const svgContent = await this.executeStageB(designJson, options)

      this.logInfo('处理完成', {
        designTemplate: designJson.template_type,
        svgLength: svgContent.length
      })

      return { svgContent, designJson }
    } catch (error) {
      this.logError('处理失败', error)
      throw error
    }
  }

  /**
   * 执行阶段A：生成设计JSON
   */
  private async executeStageA(text: string, options?: GenerationOptions): Promise<DesignJSON> {
    this.logInfo('开始阶段A - 设计分析')

    const messages: AIMessage[] = [
      { role: 'system', content: DEEPSEEK_STAGE_A_SYSTEM },
      { role: 'user', content: createStageAUserPrompt(text, { styleChoice: options?.styleChoice }) }
    ]

    const jsonResponse = await this.callAPI(messages)
    this.logInfo('阶段A 响应获取成功', { responseLength: jsonResponse.length })

    const cleanJsonResponse = this.cleanJsonResponse(jsonResponse)
    const designJson = this.parseJsonResponse<DesignJSON>(cleanJsonResponse)

    // 验证设计JSON的基本结构
    if (!designJson.template_type || !designJson.palette) {
      this.logError('设计JSON缺少必要字段', designJson)
      throw new Error(`${ERROR_MESSAGES.INVALID_JSON}: 缺少必要字段`)
    }

    this.logInfo('阶段A 完成', { templateType: designJson.template_type })
    return designJson
  }

  /**
   * 执行阶段B：生成SVG
   */
  private async executeStageB(designJson: DesignJSON, options?: GenerationOptions): Promise<string> {
    this.logInfo('开始阶段B - SVG渲染')

    const messages: AIMessage[] = [
      { role: 'system', content: DEEPSEEK_STAGE_B_SYSTEM },
      { role: 'user', content: createStageBUserPrompt(JSON.stringify(designJson), options?.styleChoice) }
    ]

    const svgResponse = await this.callAPI(messages)
    const svgContent = this.extractSvgContent(svgResponse)

    this.logInfo('阶段B 完成', { svgLength: svgContent.length })
    return svgContent
  }
}

File: pyapp/src/pyapp/services/ai/deepseek.py
import json
import httpx
from typing import Any

from pyapp.core.config import settings, logger
from pyapp.core.exceptions import AIServiceError
from pyapp.domain.models import DesignJSON, GenerationOptions
from pyapp.services.ai.base import AIService
from pyapp.services.prompts import (
    STAGE_A_SYSTEM_PROMPT,
    STAGE_B_SYSTEM_PROMPT,
    create_stage_a_user_prompt,
    create_stage_b_user_prompt,
)


class DeepSeekService(AIService):
    """DeepSeek AI Service Implementation"""

    def __init__(self):
        super().__init__("DeepSeek")
        self.api_key = settings.DEEPSEEK_API_KEY
        self.api_url = settings.DEEPSEEK_API_URL

    async def _call_api(self, messages: list[dict[str, str]]) -> str:
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }
        data = {
            "model": "deepseek-chat",
            "messages": messages,
            "temperature": 1.3, # Creative mode
            "stream": False
        }

        try:
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    self.api_url, headers=headers, json=data
                )
                
                if response.status_code != 200:
                    error_msg = f"DeepSeek API Error: {response.status_code} - {response.text}"
                    logger.error(error_msg)
                    raise AIServiceError(error_msg)
                
                result = response.json()
                return result["choices"][0]["message"]["content"]
                
        except httpx.RequestError as e:
            logger.error(f"DeepSeek Network Error: {e}")
            raise AIServiceError(f"Network error connecting to DeepSeek: {e}")
        except Exception as e:
            logger.error(f"DeepSeek Unexpected Error: {e}")
            raise AIServiceError(f"Unexpected error in DeepSeek service: {e}")

    async def process(self, text: str, options: GenerationOptions) -> tuple[str, DesignJSON]:
        logger.info(f"Starting DeepSeek processing for text length: {len(text)}")
        
        # Stage A: Analysis & Design
        stage_a_prompt = create_stage_a_user_prompt(text, options.model_dump())
        messages_a = [
            {"role": "system", "content": STAGE_A_SYSTEM_PROMPT},
            {"role": "user", "content": stage_a_prompt},
        ]
        
        try:
            raw_json = await self._call_api(messages_a)
            clean_json = self.clean_json_response(raw_json)
            design_data = json.loads(clean_json)
            design_json = DesignJSON(**design_data)
            logger.info(f"Stage A complete. Template: {design_json.template_type}")
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse Design JSON: {e}. Content: {raw_json}")
            raise AIServiceError("Failed to generate valid Design JSON")
        except Exception as e:
             raise AIServiceError(f"Stage A failed: {e}")

        # Stage B: SVG Rendering
        stage_b_prompt = create_stage_b_user_prompt(
            clean_json, options.styleChoice
        )
        messages_b = [
            {"role": "system", "content": STAGE_B_SYSTEM_PROMPT},
            {"role": "user", "content": stage_b_prompt},
        ]
        
        try:
            raw_svg = await self._call_api(messages_b)
            svg_content = self.extract_svg_content(raw_svg)
            
            if not svg_content or "<svg" not in svg_content:
                raise AIServiceError("No valid SVG content generated")
                
            logger.info(f"Stage B complete. SVG length: {len(svg_content)}")
            return svg_content, design_json
            
        except Exception as e:
            raise AIServiceError(f"Stage B failed: {e}")
